src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java

```diff
@@ -1,99 +1,256 @@
 
     /**
-     * Constructor and do collection work.
-     * @param filteredReads          reads that pass filters
-     * @param samRgRecords           a list of read groups
-     * @param HistogramMADTolerance  MAD tolerance when producing histogram plot
+     * Delegates actual work to utility functions.
+     *
+     * @param filteredReads         reads that pass filters
+     * @param header                header in the input
+     * @param accumLevels           accumulation level {ALL_READS, SAMPLE, LIBRARY, READ_GROUP}
+     * @param histogramMADTolerance MAD tolerance when producing histogram plot
      */
-    // TODO: * @param accumLevels            accumulation level {SAMPLE, LIBRARY, READGROUP}
     public InsertSizeMetricsCollectorSpark(final JavaRDD<GATKRead> filteredReads,
-                                           //final Set<MetricAccumulationLevel> accumLevels,
-                                           final List<SAMReadGroupRecord> samRgRecords,
-                                           final double HistogramMADTolerance) {
-
-        final JavaRDD<Integer> lengthRDD = filteredReads.map(read -> Math.abs(read.getFragmentLength()) );
-        lengthRDD.cache();
-
-        metrics.READ_PAIRS = filteredReads.count();
-
-        collectLevelInfo();
+                                           final SAMFileHeader header,
+                                           final Set<MetricAccumulationLevel> accumLevels,
+                                           final double histogramMADTolerance) {
+
+        // construct untrimmed hand rolled "histogram" (SortedMap) in three steps, because htsjdk Histogram does not play well with Serialization
+        // so first hand roll a histogram using Spark (read traversal is the bottleneck in terms of performance),
+        //   when computing corresponding metrics locally, use htsjdk Histogram converted the sorted map version to do the work
+        final Map<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> unsortedHistogramsAtRGLevel = filteredReads.mapToPair(read -> step1ReadTraversal(read, header))
+                                                                                                                                    .groupByKey()
+                                                                                                                                    .mapToPair(InsertSizeMetricsCollectorSpark::step2ConvertToList)
+                                                                                                                                    .mapToPair(InsertSizeMetricsCollectorSpark::step3ConvertToHistogram)
+                                                                                                                                    .collectAsMap();
+
+        final SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> histogramsAtRGLevel = new TreeMap<>(groupInfoComparator);
```

Comment:
I'm not sure it makes sense to have groupInfoComparator be a static variable on the class. Why not just new() one here?